cudax/include/cuda/experimental/__memory_resource/cuda_async_memory_resource.h

```diff
@@ -0,0 +1,294 @@
+        && (_CUDA_VMR::default_cuda_malloc_alignment % __alignment == 0);
+  }
+
+  //! @brief  Returns the default ``cudaMemPool_t`` from the specified device.
+  //! @throws cuda_error if retrieving the default ``cudaMemPool_t`` fails.
+  //! @returns The default memory pool of the specified device.
+  _CCCL_NODISCARD static ::cudaMemPool_t __get_default_mem_pool(const int __device_id)
+  {
+    _LIBCUDACXX_ASSERT(_CUDA_VMR::__device_supports_pools(__device_id), "cudaMallocAsync not supported");
+
+    ::cudaMemPool_t __pool;
+    _CCCL_TRY_CUDA_API(
+      ::cudaDeviceGetDefaultMemPool, "Failed to call cudaDeviceGetDefaultMemPool", &__pool, __device_id);
+    return __pool;
+  }
+
+public:
+  //! @brief Default constructs the cuda_async_memory_resource using the default \c cudaMemPool_t of the default device.
+  //! @throws cuda_error if retrieving the default \c cudaMemPool_t fails.
+  cuda_async_memory_resource()
+      : __pool_(__get_default_mem_pool(0))
+  {}
+
+  //! @brief Constructs a cuda_async_memory_resource using the default \c cudaMemPool_t of a given device.
+  //! @throws cuda_error if retrieving the default \c cudaMemPool_t fails.
+  cuda_async_memory_resource(::cuda::experimental::device_ref __device)
+      : __pool_(__get_default_mem_pool(__device.get()))
+  {}
+
+  cuda_async_memory_resource(int)                   = delete;
+  cuda_async_memory_resource(_CUDA_VSTD::nullptr_t) = delete;
+
+  //! @brief  Constructs the cuda_async_memory_resource from a \c cudaMemPool_t.
+  //! @param __pool The \c cudaMemPool_t used to allocate memory.
+  explicit cuda_async_memory_resource(::cudaMemPool_t __pool) noexcept
+      : __pool_(__pool)
+  {}
+
+  //! @brief  Constructs the cuda_async_memory_resource from a \c cuda_memory_pool by calling native_handle().
+  //! @param __pool The \c cuda_memory_pool used to allocate memory.
+  explicit cuda_async_memory_resource(cuda_memory_pool& __pool) noexcept
+      : __pool_(__pool.native_handle())
+  {}
+
+  //! @brief Allocate device memory of size at least \p __bytes via cudaMallocFromPoolAsync.
+  //! @param __bytes The size in bytes of the allocation.
+  //! @param __alignment The requested alignment of the allocation.
+  //! @throws cuda::std::bad_alloc In case of invalid alignment.
+  //! @throws cuda::cuda_error If an error code was return by the CUDA API call.
+  //! @returns Pointer to the newly allocated memory.
+  _CCCL_NODISCARD void* allocate(const size_t __bytes,
+                                 const size_t __alignment = _CUDA_VMR::default_cuda_malloc_alignment)
+  {
+    if (!__is_valid_alignment(__alignment))
+    {
+      _CUDA_VSTD_NOVERSION::__throw_bad_alloc();
+    }
+
+    void* __ptr{nullptr};
+    _CCCL_TRY_CUDA_API(
+      ::cudaMallocFromPoolAsync,
+      "cuda_async_memory_resource::allocate failed to allocate with cudaMallocFromPoolAsync",
+      &__ptr,
+      __bytes,
+      __pool_,
+      ::cudaStream_t{0});
```

Comment:
I can add an internal stream to the class so that we always have a unique one

Reply:
It turns out this is a bit trickier than expected, we need to ensure that we construct the stream with the right device and we do not have that available.

Also I dont think the example is a problem, because we require the memory to be `__device__` only

Reply:
right, only `__device__` memory can be allocated asynchronously -- for now. i have been told that async allocation of managed memory will be supported next year. i expect such an API to require a stream.

Reply:
Even for device memory I think doing a synchronous allocation like this is a problem. Especially since `cudax::stream` is non-blocking by default. In a sequence of `auto ptr = mr.allocate()` -> `launch(stream, ..., ptr)` user would expect synchronous allocation to be usable right away, but with stream being non-blocking by default this will fail most likely.

We would need to change the semantics (how?) or change the implementation to synchronize the stream the allocation is done on.